snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -133,6 +133,97 @@ def _update_yaml_with_extracted_metadata(
         raise meta_errors.MissingSnapcraftYamlKeysError(keys=missing_keys)
 
 
+def _adopt_info(
+        config_data: Dict[str, Any],
+        extracted_metadata: _metadata.ExtractedMetadata):
+    metadata_dict = extracted_metadata.to_dict()
+    for key, value in filter(
+            lambda item: item != 'desktop',
+            metadata_dict.items()):
+        if key not in config_data:
+            if key == 'icon':
+                if _icon_file_exists() or not os.path.exists(value):
+                    # Do not overwrite the icon file.
+                    continue
```

Comment:
My understanding of this is: "If the icon has been specified by its conventional location, or the icon specified in the extracted metadata does not exist, skip over it."

Assuming my understanding is correct, I of course agree with giving the locally-supplied icon priority, but if the icon specified in the extracted metadata doesn't exist, I kinda feel like that should be an error. Is there a reason it's not?

Reply:
I was thinking about cases when the icon is moved during a step in the build phase, or the packaging phase. I think that's common. So, I was imagining a case where we find the appstream file, but the icon is not in the location specified there because that's a packaging build task that doesn't run on the normal compilation.
Maybe I was thinking too much? Should we wait for people to complain about us forcing them to move the icon around?

Reply:
No actually, I think you're right. Since we check several different root directories for this file, it may very well not exist.